File: cpp-test/oracle/height.c
/*
    Copyright © 2026 Mikhail Hogrefe

    This file is part of Malachite.

    Malachite is free software: you can redistribute it and/or modify it under the terms of the
    GNU Lesser General Public License (LGPL) as published by the Free Software Foundation; either
    version 3 of the License, or (at your option) any later version. See
    <https://www.gnu.org/licenses/>.

    Diffs `x.to_height() = h` and `x.into_height() = h` lines against fmpq_height, and
    `x.height_significant_bits() = b` lines against fmpq_height_bits. The rational receiver
    prints as `p/q` or `p`, which fmpq_set_str reads directly.
*/

#include <stdlib.h>
#include <string.h>

#include <flint/fmpq.h>

#include "oracle.h"

static int
check_height_value_line(char * line, int line_number, const char * needle)
{
    char * start = strstr(line, needle);
    if (start == NULL)
    {
        return 0;
    }
    *start = '\0';
    char * expected_str = start + strlen(needle);
    int result = 0;
    fmpq_t x;
    fmpz_t height, expected;
    fmpq_init(x);
    fmpz_init(height);
    fmpz_init(expected);
    if (fmpq_set_str(x, line, 10) != 0)
    {
        flint_printf("error in fmpq_height test, line %d: unreadable rational\n", line_number);
        result = 1;
    }
    else
    {
        fmpz_set_str(expected, expected_str, 10);
        fmpq_height(height, x);
        if (!fmpz_equal(height, expected))
        {
            flint_printf("error in fmpq_height test, line %d. FLINT: ", line_number);
            fmpz_print(height);
            flint_printf("\n");
            result = 1;
        }
    }
    fmpq_clear(x);
    fmpz_clear(height);
    fmpz_clear(expected);
    return result;
}

static int
check_to_height_line(char * line, int line_number)
{
    return check_height_value_line(line, line_number, ".to_height() = ");
}

static int
check_into_height_line(char * line, int line_number)
{
    return check_height_value_line(line, line_number, ".into_height() = ");
}

static int
check_height_bits_line(char * line, int line_number)
{
    char * start = strstr(line, ".height_significant_bits() = ");
    if (start == NULL)
    {
        return 0;
    }
    *start = '\0';
    ulong expected = strtoull(start + strlen(".height_significant_bits() = "), NULL, 10);
    int result = 0;
    fmpq_t x;
    fmpq_init(x);
    if (fmpq_set_str(x, line, 10) != 0)
    {
        flint_printf("error in fmpq_height_bits test, line %d: unreadable rational\n",
                     line_number);
        result = 1;
    }
    else if (fmpq_height_bits(x) != expected)
    {
        flint_printf("error in fmpq_height_bits test, line %d. FLINT: %wu\n", line_number,
                     fmpq_height_bits(x));
        result = 1;
    }
    fmpq_clear(x);
    return result;
}

int
run_fmpq_height(const char * arg)
{
    int r = for_each_line(arg, check_to_height_line);
    if (r != 0)
    {
        return r;
    }
    return for_each_line(arg, check_into_height_line);
}

int
run_fmpq_height_bits(const char * arg)
{
    return for_each_line(arg, check_height_bits_line);
}

File: malachite-q/src/bin_util/demo_and_bench/rational/arithmetic/height.rs
// Copyright © 2026 Mikhail Hogrefe
//
// This file is part of Malachite.
//
// Malachite is free software: you can redistribute it and/or modify it under the terms of the GNU
// Lesser General Public License (LGPL) as published by the Free Software Foundation; either version
// 3 of the License, or (at your option) any later version. See <https://www.gnu.org/licenses/>.

use malachite_base::test_util::bench::{BenchmarkType, run_benchmark};
use malachite_base::test_util::generators::common::{GenConfig, GenMode};
use malachite_base::test_util::runner::Runner;
use malachite_q::test_util::bench::bucketers::rational_bit_bucketer;
use malachite_q::test_util::generators::rational_gen;

pub(crate) fn register(runner: &mut Runner) {
    register_demo!(runner, demo_rational_to_height);
    register_demo!(runner, demo_rational_into_height);
    register_demo!(runner, demo_rational_height_significant_bits);

    register_bench!(runner, benchmark_rational_height_evaluation_strategy);
    register_bench!(runner, benchmark_rational_height_significant_bits);
}

fn demo_rational_to_height(gm: GenMode, config: &GenConfig, limit: usize) {
    for x in rational_gen().get(gm, config).take(limit) {
        println!("{x}.to_height() = {}", x.to_height());
    }
}

fn demo_rational_into_height(gm: GenMode, config: &GenConfig, limit: usize) {
    for x in rational_gen().get(gm, config).take(limit) {
        let x_old = x.clone();
        println!("{x_old}.into_height() = {}", x.into_height());
    }
}

fn demo_rational_height_significant_bits(gm: GenMode, config: &GenConfig, limit: usize) {
    for x in rational_gen().get(gm, config).take(limit) {
        println!(
            "{x}.height_significant_bits() = {}",
            x.height_significant_bits()
        );
    }
}

fn benchmark_rational_height_evaluation_strategy(
    gm: GenMode,
    config: &GenConfig,
    limit: usize,
    file_name: &str,
) {
    run_benchmark(
        "Rational.to_height()",
        BenchmarkType::EvaluationStrategy,
        rational_gen().get(gm, config),
        gm.name(),
        limit,
        file_name,
        &rational_bit_bucketer("x"),
        &mut [
            ("Rational.to_height()", &mut |x| {
                no_out!(x.to_height());
            }),
            ("Rational.into_height()", &mut |x| {
                no_out!(x.into_height());
            }),
        ],
    );
}

fn benchmark_rational_height_significant_bits(
    gm: GenMode,
    config: &GenConfig,
    limit: usize,
    file_name: &str,
) {
    run_benchmark(
        "Rational.height_significant_bits()",
        BenchmarkType::Single,
        rational_gen().get(gm, config),
        gm.name(),
        limit,
        file_name,
        &rational_bit_bucketer("x"),
        &mut [("Malachite", &mut |x| {
            no_out!(x.height_significant_bits());
        })],
    );
}
